File: heart/src/main/java/com/vantagecircle/heartrate/processing/Processing.java
package com.vantagecircle.heartrate.processing;

import android.media.Image;
import android.os.Build;
import android.support.annotation.RequiresApi;
import android.util.Log;

import com.vantagecircle.heartrate.core.HeartRate;

import java.nio.ByteBuffer;

public class Processing implements ProcessingSupport {
    private static final String TAG = Processing.class.getSimpleName();

    @Override
    @RequiresApi(api = Build.VERSION_CODES.KITKAT)
    public byte[] YUV_420_888toNV21(Image image) {
        byte[] nv21;
        ByteBuffer yBuffer = image.getPlanes()[0].getBuffer();
        ByteBuffer uBuffer = image.getPlanes()[1].getBuffer();
        ByteBuffer vBuffer = image.getPlanes()[2].getBuffer();

        int ySize = yBuffer.remaining();
        int uSize = uBuffer.remaining();
        int vSize = vBuffer.remaining();

        nv21 = new byte[ySize + uSize + vSize];

        //U and V are swapped
        yBuffer.get(nv21, 0, ySize);
        vBuffer.get(nv21, ySize, vSize);
        uBuffer.get(nv21, ySize + vSize, uSize);

        return nv21;
    }

    @Override
    public int YUV420SPtoRedAvg(byte[] yuv420sp, int width, int height) {
        if (yuv420sp == null) return 0;
        final int frameSize = width * height;
        int sum = YUV420SPtoRedSum(yuv420sp, width, height);
        //Log.e(TAG, "Red Avg " + sum / frameSize);
        return (sum / frameSize);
    }

    private int YUV420SPtoRedSum(byte[] yuv420sp, int width, int height) {
        if (yuv420sp == null) return 0;
        final int frameSize = width * height;
        int sum = 0;
        for (int j = 0, yp = 0; j < height; j++) {
            int uvp = frameSize + (j >> 1) * width, u = 0, v = 0;
            for (int i = 0; i < width; i++, yp++) {
                int y = (0xff & yuv420sp[yp]) - 16;
                if (y < 0) y = 0;
                if ((i & 1) == 0) {
                    v = (0xff & yuv420sp[uvp++]) - 128;
                    u = (0xff & yuv420sp[uvp++]) - 128;
                }
                int y1192 = 1192 * y;
                int r = (y1192 + 1634 * v);
                int g = (y1192 - 833 * v - 400 * u);
                int b = (y1192 + 2066 * u);
                if (r < 0)
                    r = 0;
                else if (r > 262143)
                    r = 262143;
                if (g < 0)
                    g = 0;
                else if (g > 262143)
                    g = 262143;
                if (b < 0)
                    b = 0;
                else if (b > 262143)
                    b = 262143;
                int pixel = 0xff000000 | ((r << 6) & 0xff0000) | ((g >> 2) & 0xff00) | ((b >> 10) & 0xff);
                int red = (pixel >> 16) & 0xff;
                sum += red;
            }
        }
        //Log.e(TAG, "Red sum " + sum);
        return sum;
    }
}

File: app/src/main/java/com/bapidas/heartrate/processing/Processing.kt
package com.bapidas.heartrate.processing

import android.media.Image
import java.nio.ByteBuffer

/**
 * Updated processing class with clearer naming and optimized YUV-to-RGB conversion.
 */
class Processing : ProcessingSupport {

    override fun toNv21(image: Image): ByteArray {
        val yBuffer = image.planes[0].buffer
        val uBuffer = image.planes[1].buffer
        val vBuffer = image.planes[2].buffer

        val ySize = yBuffer.remaining()
        val uSize = uBuffer.remaining()
        val vSize = vBuffer.remaining()

        val nv21 = ByteArray(ySize + uSize + vSize)

        // NV21 is YYYY VUVU. Note: U and V are often swapped in internal buffers.
        yBuffer.get(nv21, 0, ySize)
        vBuffer.get(nv21, ySize, vSize)
        uBuffer.get(nv21, ySize + vSize, uSize)

        return nv21
    }

    override fun calculateAverageRed(nv21: ByteArray, width: Int, height: Int): Int {
        val frameSize = width * height
        var redSum = 0L // Use Long to avoid overflow during summation
        var yp = 0

        for (j in 0 until height) {
            var uvp = frameSize + (j shr 1) * width
            var u = 0
            var v = 0
            
            for (i in 0 until width) {
                // Extract Y component
                val yRaw = (0xff and nv21[yp].toInt()) - 16
                val y = if (yRaw < 0) 0 else yRaw
                
                // Extract U and V components every 2 pixels (since it's 4:2:0)
                if (i and 1 == 0) {
                    v = (0xff and nv21[uvp++].toInt()) - 128
                    u = (0xff and nv21[uvp++].toInt()) - 128
                }
                
                // standard YUV to RGB conversion formula for Red
                val y1192 = 1192 * y
                val rRaw = y1192 + 1634 * v
                
                // Clamp R value between 0 and 262143
                val r = when {
                    rRaw < 0 -> 0
                    rRaw > 262143 -> 262143
                    else -> rRaw
                }
                
                // Extract the red byte (scaled back to 0-255 range)
                val red = (r shr 10) and 0xff
                redSum += red
                
                yp++
            }
        }
        
        return if (frameSize > 0) (redSum / frameSize).toInt() else 0
    }

    companion object {
        private val TAG = Processing::class.java.simpleName
    }
}
